### PR TITLE
feat: add `splice` function

### PR DESCRIPTION
**Pull request by fneco**
I would like to add  a `splice` function that works almost identically to the `Array.splice` method but does not mutate the data and does not return a deleted array.

Thanks for creating a great library!

---

Make sure that you:

- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`
- [x] New methods added to `mapping.md`

---

<details><summary>We use semantic PR titles to automate the release process!</summary>

https://conventionalcommits.org

PRs should be titled following using the format: `< TYPE >(< scope >)?: description`

### Available Types:

- `feat`: new functions!
- `fix`: changes to function implementations that fix a bug where a function produces the wrong _runtime_ results.
- `type`: type-only changes (params, overloading, return types, etc...)
- `perf`: changes to function implementations that improve a functions _runtime_ performance.
- `refactor`: changes to function implementations that are neither `fix` nor `perf`
- `test`: tests-only changes (transparent to users of the function).
- `docs`: changes to the documentation of a function **or the documentation site**.
- `build`, `ci`, `style`, `chore`, and `revert`: are only relevant for the internals of the library.

For scope put the name of the function you are working on (either new or
existing).

</details>

### CODE REVIEW

**Review by eranhirsch — CHANGES_REQUESTED** [review]
So this is just a non-mutating `splice` with a purried data-last impl. I think we should name it `splice`. Except people coming from Ramda, people who are more familiar with vanilla JS wouldn't except it to be named differently.

Once renamed, we should probably also accept the additional "replacement" array (but not as a variadic param as that would break purrying) so that it is complete.

**Comment by fneco** [comment]
Thanks for checking my PR and replying!
I renamed `remove` to `splice` and add `replacement` parameter.  9e51ff7
And I editted PR title and body.

File: src/splice.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * Removes elements from an array and, inserts new elements in their place.
+ * @param items the array to splice.
+ * @param start the zero-based location in the array from which to start removing elements.
+ * @param deleteCount the number of elements to remove.
+ * @param replacement the elements to insert into the items in place of the deleted elements.
+ * @signature
+ *    R.splice(items, start, deleteCount, replacement)
+ * @example
+ *    R.splice(2, 3, [1,2,3,4,5,6,7,8], []); //=> [1,2,6,7,8]
+ *    R.splice(2, 3, [1,2,3,4,5,6,7,8], [9, 10]); //=> [1,2,9,10,6,7,8]
```
Comment (eranhirsch) [inline src/splice.ts:13]: These examples are wrong and don't match the signature...
```suggestion
 *    R.splice([1,2,3,4,5,6,7,8], 2, 3, []); //=> [1,2,6,7,8]
 *    R.splice([1,2,3,4,5,6,7,8], 2, 3, [9, 10]); //=> [1,2,9,10,6,7,8]
```

File: src/splice.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Removes elements from an array and, inserts new elements in their place.
+ * @param items the array to splice.
+ * @param start the zero-based location in the array from which to start removing elements.
+ * @param deleteCount the number of elements to remove.
+ * @param replacement the elements to insert into the items in place of the deleted elements.
+ * @signature
+ *    R.splice(start, deleteCount, replacement)(items)
+ * @example
+ *    R.pipe([1,2,3,4,5,6,7,8], R.splice(2, 3)) // => [1,2,6,7,8]
```
Comment (eranhirsch) [inline src/splice.ts:33]: This example is wrong, the replacement array is required, it should always be provided (empty here)
```suggestion
 *    R.pipe([1,2,3,4,5,6,7,8], R.splice(2, 3, [])) // => [1,2,6,7,8]
```

File: src/splice.test.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    expect(splice([1, 2, 3, 4, 5, 6, 7, 8], 2, 3, [9, 10])).toEqual([
+      1, 2, 9, 10, 6, 7, 8,
+    ]);
+  });
+
+  test('immutability', () => {
+    const data = [1, 2, 3];
+    const result = splice(data, 0, 0, []);
+    expect(result).toEqual(data);
+    assert.notStrictEqual(result, data);
```
Comment (eranhirsch) [inline src/splice.test.ts:16]: I think it's easier when all assertions are of the same flavor, we usually use expect so you should use it here too
```suggestion
    expect(result).not.toBe(data);
```

File: src/splice.test.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+describe('data first', () => {
+  test('splice', () => {
+    expect(splice([1, 2, 3, 4, 5, 6, 7, 8], 2, 3, [])).toEqual([1, 2, 6, 7, 8]);
+    expect(splice([1, 2, 3, 4, 5, 6, 7, 8], 2, 3, [9, 10])).toEqual([
+      1, 2, 9, 10, 6, 7, 8,
+    ]);
+  });
+
+  test('immutability', () => {
```
Comment (eranhirsch) [inline src/splice.test.ts:12]: cool test

**Comment by TkDodo** [comment]
@eranhirsch I'm not sure if we should have a function named splice that is always associated with mutating its input 🤔

File: src/splice.test.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { pipe } from './pipe';
+import { splice } from './splice';
+
+describe('data first', () => {
+  test('splice', () => {
+    expect(splice([1, 2, 3, 4, 5, 6, 7, 8], 2, 3, [])).toEqual([1, 2, 6, 7, 8]);
+    expect(splice([1, 2, 3, 4, 5, 6, 7, 8], 2, 3, [9, 10])).toEqual([
+      1, 2, 9, 10, 6, 7, 8,
+    ]);
```
Comment (eranhirsch) [inline src/splice.test.ts:9]: I'd split this into several tests. There's more value in having a unique test for each case (here: with replacement and without).

**Review by eranhirsch — CHANGES_REQUESTED** [review]
This is great, just a few nits. Can you also add a few more tests to cover all the edge cases (index > length, index < 0, deleteCount > length, deleteCount < 0, etc...) so that if anyone changes the impl, they will know if they are changing an output that someone might be expecting.

**Comment by eranhirsch** [comment]
@TkDodo , it's called splice even in the non mutating version. But the toXXX naming convention wouldn't work for us

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSpliced

File: src/splice.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import { purry } from './purry';
+
+/**
+ * Removes elements from an array and, inserts new elements in their place.
+ * @param items the array to splice.
+ * @param start the zero-based location in the array from which to start removing elements.
+ * @param deleteCount the number of elements to remove.
+ * @param replacement the elements to insert into the items in place of the deleted elements.
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/splice.ts]: I corrected it in 6083c89 .

File: src/splice.ts (hunk 1 above)
Comment (fneco) [reply src/splice.ts]: Sorry, I corrected it in 6083c89 .

File: src/splice.test.ts (hunk 3 above)
Comment (fneco) [reply src/splice.test.ts]: I corrected it in this line: https://github.com/remeda/remeda/pull/393/files#diff-0696fb3eb6df2fdeb8e06f4dab9cd586776e32d60d7001dc8eb02cd00ad58061R23

File: src/splice.test.ts (hunk 5 above)
Comment (fneco) [reply src/splice.test.ts]: I fixed test case and assertion to be one-to-one. 

> This is great, just a few nits. Can you also add a few more tests to cover all the edge cases (index > length, index < 0, deleteCount > length, deleteCount < 0, etc...) so that if anyone changes the impl, they will know if they are changing an output that someone might be expecting.

I added regression testing. I tested with matrix data to include as many edge cases as possible.

cf. https://github.com/remeda/remeda/pull/393/files#diff-0696fb3eb6df2fdeb8e06f4dab9cd586776e32d60d7001dc8eb02cd00ad58061

**Review by eranhirsch — APPROVED** [review]
Thanks for this addition to the library, it's great
